### ISSUE TITLE
Correctly type hinted methods do not show up in the method picker when connecting signals with type hints for custom classes.

### ISSUE THREAD

**Issue by sday-ssc** [godotengine/godot#84058]
### Godot version

4.2.beta3

### System information

Windows 11

### Issue description

The method picker for connecting signals in the editor will filter out correctly type hinted methods if the signal's arguments have a custom class type hint. It only shows up if you disable "Compatible Methods Only".

### Steps to reproduce

Write a custom class, like "CustomData".
If you write this signal on a node:
  signal custom_data_updated (custom_data: CustomData)

And then write this method on another node:
  on_data_updated(custom_data: CustomData) -> void:

That method will not show up when picking a method from that node to connect to the signal. It only shows up if you turn off "Compatible Methods Only".

### Minimal reproduction project

[CustomClassSignal.zip](https://github.com/godotengine/godot/files/13190214/CustomClassSignal.zip)

**Comment by RicardRC** [comment]
I'm also experiencing this. Not a huge problem, but it does waste dev time doubting oneself and checking around until you find out it's a bug.

**Comment by granitrocky** [comment]
@akien-mga I don't think this was actually completed for 4.3. The PR you linked says explicitly:
"Also a side note: this PR does NOT fixes https://github.com/godotengine/godot/issues/84058"

I think this should be reopened

**Comment by HolonProduction** [comment]
Issue is here:

https://github.com/HolonProduction/godot/blob/ccb8004960b1638f5532048db6c33a0bf5f0769e/editor/scene/connections_dialog.cpp#L358

`ClassDB` is not responsible for global script classes and can't give correct results on their inheritance.

**Comment by Francis-of-ASCII** [comment]
Hello! Is this issue still available? I have reproduced it on v4.7.1

**Comment by HolonProduction** [comment]
Yes, the issue is still valid.

**Comment by Francis-of-ASCII** [comment]
Like @HolonProduction said, `ClassDB` will not give correct results on a global script classes' inheritance. That means `is_parent_class` won't work on the line @HolonProduction pointed out. I think I have found a solution: Using `get_global_class_list` from `ProjectSettings`, get the base class of the custom class. If it came from another custom class, get the base of that class, until it is a class in `ClassDB`. We can then use `is_parent_class` from `ClassDB` to check whether the base class has an ancestor of `inherits`. 

Wrote this GDScript function to demonstrate finding the base:
```
func get_base_of_custom_class(custom_class: StringName) -> StringName:
	var base_class: StringName
	var idx = -1
	
	# Go through custom classes to find custom_class
	for i in range(len(ProjectSettings.get_global_class_list())):
		if ProjectSettings.get_global_class_list()[i]["class"] == custom_class:
			idx = i
			
	if idx != -1:
		
		# Find where the custom class came from
		base_class = ProjectSettings.get_global_class_list()[idx]["base"]
			
		# If it came from another custom class, get the base of that class
		while !ClassDB.get_class_list().has(base_class):
			for o in range(len(ProjectSettings.get_global_class_list())):
				if ProjectSettings.get_global_class_list()[o]["class"] == base_class:
					base_class = ProjectSettings.get_global_class_list()[o]["base"]
					
		return base_class
	else:
		# The class was not a custom_class
		return custom_class
```
Then just pass the results of this into `is_parent_class` from `ClassDB`. Will rewrite this in C++ if it looks good.

**Comment by HolonProduction** [comment]
Logic lile this is already implemented in `ScriptServer`. I can't give you the correct method of the top of my head, but it's somewhere in there.

**Comment by Francis-of-ASCII** [comment]
I have patched the issue using a slightly different method.

`type_mismatch` should be true when the signal class is not the same as the method class, or the method class is a descendant of the signal class. For the first one, (the signal class is not the same as the method class), I patched it by adding `effective_args[i].second != mi.arguments[i].class_name` to the if statement on the line of the issue. For the second, (the method class is a descendant of the signal class), I patched it by first rewriting it. It is the same as saying that the method class inherits the signal class. Use `get_inheriter_list` from `ScriptServer` and check whether the signal class is in the list. If it is, then `type_mismatch` is true.

Completed, the if statement looks like this:

```
if (stype == Variant::OBJECT && mtype == Variant::OBJECT && !ClassDB::is_parent_class(effective_args[i].second, mi.arguments[i].class_name) && effective_args[i].second != mi.arguments[i].class_name) {

					if (ScriptServer::is_global_class(effective_args[i].second) == true) {
						List<StringName> inheritors;
						ScriptServer::get_inheriters_list(effective_args[i].second, &inheritors);
						for (const StringName &S : inheritors) {
							if (S == mi.arguments[i].class_name) {
								type_mismatch = true;

								break;
							}
						}
						if (type_mismatch == true) {
							break;
						}
					} else {
						type_mismatch = true;
					
						break;
					}
				}
```

I built it, and it works. If it looks good, I can write a test and create a PR if that sounds good, @HolonProduction.

**Comment by HolonProduction** [comment]
Please use `EditorData::script_class_is_parent` and `ScriptServer::is_global_class` to determine whether to call `script_class_is_parent` or `ClassDB::is_parent_class`. Keep in mind that class db must never receive the name of a global class and `script_class_is_parent` must be passed the name of a global class for it's first parameter.

I think this are all the building blocks to get a minimal fix for this issue together.

**Comment by Francis-of-ASCII** [comment]
Made the changes, works well. Does this look good?

```
				if (stype == Variant::OBJECT && mtype == Variant::OBJECT && effective_args[i].second != mi.arguments[i].class_name) {
					if (ScriptServer::is_global_class(effective_args[i].second) == true || ScriptServer::is_global_class(mi.arguments[i].class_name) == true) {
						if (EditorNode::get_editor_data().script_class_is_parent(effective_args[i].second, mi.arguments[i].class_name) == false) {
							type_mismatch = true;
							break;
						}
					} else if (ClassDB::is_parent_class(effective_args[i].second, mi.arguments[i].class_name) == false) {
						type_mismatch = true;
						break;
					}
				}
```

**Comment by Francis-of-ASCII** [comment]
As far as I know, there isn't already a test written for connections_dialog.cpp at all. @HolonProduction would you suggest writing a new test for this issue?

**Comment by HolonProduction** [comment]
If you have a working fix, I'd suggest you open a PR. That way the relevant team can take a look. I don't know what the requirements around testing are in the editor area.
